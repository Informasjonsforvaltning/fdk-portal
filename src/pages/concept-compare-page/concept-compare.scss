@import '../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';


.scrollable {
  overflow: auto;

  .sticky th {
    background-color: $color-neutral-lightest;
    border-bottom: 0;
    border-top: 0;
    position: sticky;
    top: 0;
    z-index: 5;
  }

  .table.scrollable {
    display: block;
  }

  .table.scrollable tbody {
    display: table;
  }

  th, td {
    padding: 1em !important;
    min-width: 16em;
  }

  td:nth-child(1) {
    background-color: $color-neutral-lightest;
    position: sticky;
    left     : 0;
    z-index  : 10;
  }

  th:nth-child(1) {
    background-color: $color-neutral-lightest;
    position: sticky;
    left     : 0;
    top      : 0;
    z-index  : 200 !important;
  }

  thead tr:nth-child(1) th {
    position: sticky;
    top: 0;
    z-index: 10;
  }
}
