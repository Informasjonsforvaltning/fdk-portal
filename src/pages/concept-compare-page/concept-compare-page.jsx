import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import qs from 'qs';
import DocumentMeta from 'react-document-meta';

import localization from '../../lib/localization';
import { getTranslateText } from '../../lib/translateText';
import { LinkExternal } from '../../components/link-external/link-external.component';
import './concept-compare.scss';

const onDeleteConcept = (id, history, conceptIdsArray, removeConcept) => {
  removeConcept(id);
  const filteredConceptIds = conceptIdsArray.filter(item => item !== id);
  history.push(`?compare=${filteredConceptIds}`);
};

const renderTitle = (label, items, field) => {
  const fields = items =>
    Object.keys(items).map((item, index) => (
      <th key={`row-title-${field}-${index}`}>
        <h3>{getTranslateText(_.get(items[item], field))}</h3>
      </th>
    ));

  return (
    <thead className="sticky">
      <tr>
        <th>{label}</th>
        {fields(items)}
      </tr>
    </thead>
  );
};

const existFieldValue = value =>
  Array.isArray(value) ? value.some(valueItem => !!valueItem) : !!value;

const existValuesOnAnyItem = (items, fieldPath) =>
  Object.values(items).some(item => existFieldValue(_.get(item, fieldPath)));

const renderFieldValue = (item, fieldPath, index) => {
  const fieldValue = _.get(item, fieldPath);
  const renderArrayItem = (value, index) => (
    <span key={index} className="mr-2">
      {getTranslateText(value)}
    </span>
  );

  return (
    <td key={`row-${fieldPath}-${index}`}>
      {Array.isArray(fieldValue)
        ? fieldValue.map(renderArrayItem)
        : getTranslateText(fieldValue)}
    </td>
  );
};

const renderRow = (label, items, fieldPath) => {
  if (!existValuesOnAnyItem(items, fieldPath)) {
    return null;
  }
  return (
    <tr>
      <td>
        <strong>{label}</strong>
      </td>
      {Object.values(items).map((item, index) =>
        renderFieldValue(item, fieldPath, index)
      )}
    </tr>
  );
};

const renderRowUrl = (label, items, fieldPath) => {
  const urlItem = (item, fieldPath, index) => {
    const fieldValue = _.get(item, fieldPath);
    return (
      <td key={`row-${fieldPath}-${index}`}>
        {_.get(fieldValue, 'uri') ? (
          <LinkExternal
            uri={_.get(fieldValue, 'uri')}
            prefLabel={
              _.get(fieldValue, 'prefLabel') || _.get(fieldValue, 'uri')
            }
          />
        ) : (
          getTranslateText(_.get(fieldValue, 'prefLabel'))
        )}
      </td>
    );
  };

  if (!existValuesOnAnyItem(items, fieldPath)) {
    return null;
  }

  return (
    <tr>
      <td>
        <strong>{label}</strong>
      </td>
      {Object.values(items).map((item, index) =>
        urlItem(item, fieldPath, index)
      )}
    </tr>
  );
};

const renderRemoveItem = (items, history, conceptIdsArray, removeConcept) => {
  const removeButtons = items =>
    Object.keys(items).map((item, index) => (
      <td key={`row-button-${index}`}>
        <button
          type="button"
          className="btn fdk-text-size-15 fdk-color-link bg-transparent"
          onClick={() => {
            onDeleteConcept(
              _.get(items, [item, 'id']),
              history,
              conceptIdsArray,
              removeConcept
            );
          }}
        >
          <i className="fa fa-minus-circle" />
          &nbsp;
          {localization.compare.removeCompare}
        </button>
      </td>
    ));

  return (
    <tr>
      <td />
      {removeButtons(items)}
    </tr>
  );
};

export const ConceptComparePage = props => {
  const {
    conceptsCompare,
    fetchConceptsToCompareIfNeeded,
    removeConcept,
    location,
    history
  } = props;
  const search = qs.parse(_.get(location, 'search'), {
    ignoreQueryPrefix: true
  });
  const conceptIdsArray = _.get(search, 'compare', '').split(',');
  fetchConceptsToCompareIfNeeded(conceptIdsArray);

  const meta = {
    title: localization.menu.conceptsCompare
  };

  return (
    <main id="content" className="container">
      <article>
        <div className="row">
          <div className="col-12">
            <DocumentMeta {...meta} />
            {conceptsCompare && (
              <>
                <h1 className="title">
                  {localization.menu.conceptsCompare} (
                  {Object.keys(conceptsCompare).length})
                </h1>

                <section className="scrollable">
                  <table className="table">
                    {renderTitle(
                      localization.facet.concept,
                      conceptsCompare,
                      'prefLabel'
                    )}
                    <tbody>
                      {renderRow(localization.responsible, conceptsCompare, [
                        'publisher',
                        'prefLabel'
                      ])}
                      {renderRow(
                        localization.compare.definition,
                        conceptsCompare,
                        ['definition', 'text']
                      )}
                      {renderRowUrl(
                        localization.compare.source,
                        conceptsCompare,
                        ['definition', 'source']
                      )}
                      {renderRow(
                        localization.compare.subject,
                        conceptsCompare,
                        'subject'
                      )}
                      {renderRow(
                        localization.compare.altLabel,
                        conceptsCompare,
                        'altLabel'
                      )}
                      {renderRow(
                        localization.compare.hiddenLabel,
                        conceptsCompare,
                        'hiddenLabel'
                      )}

                      {renderRemoveItem(
                        conceptsCompare,
                        history,
                        conceptIdsArray,
                        removeConcept
                      )}
                    </tbody>
                  </table>
                </section>
              </>
            )}
          </div>
        </div>
      </article>
    </main>
  );
};

ConceptComparePage.defaultProps = {
  conceptsCompare: null,
  fetchConceptsToCompareIfNeeded: _.noop
};

ConceptComparePage.propTypes = {
  fetchConceptsToCompareIfNeeded: PropTypes.func,
  conceptsCompare: PropTypes.object
};
