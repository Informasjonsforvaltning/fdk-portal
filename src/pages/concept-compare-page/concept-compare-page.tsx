import React, { FC, useEffect } from 'react';
import { useLocation, useHistory } from 'react-router-dom';
import _ from 'lodash';
import qs from 'qs';
import { Helmet } from 'react-helmet';

import LinkExternal from '@fellesdatakatalog/link';

import localization from '../../lib/localization';
import { getTranslateText } from '../../lib/translateText';

import './concept-compare.scss';

interface Props {
  fetchConceptsToCompareIfNeeded?: (ids: string[]) => void;
  removeConcept?: (id: string) => void;
  conceptsCompare?: Record<string, any>;
}

const onDeleteConcept = (
  id: string,
  history: any,
  conceptIdsArray: any[],
  removeConcept: any
) => {
  removeConcept(id);
  const filteredConceptIds = conceptIdsArray.filter(item => item !== id);
  history.push(`?compare=${filteredConceptIds}`);
};

const renderTitle = (label: any, items: any, field: any) => (
  <thead className='sticky'>
    <tr>
      <th>{label}</th>
      {Object.keys(items).map((item, index) => (
        <th key={`row-title-${field}-${index}`}>
          <h3>{getTranslateText(_.get(items[item], field))}</h3>
        </th>
      ))}
    </tr>
  </thead>
);

const existFieldValue = (value: any) =>
  Array.isArray(value) ? value.some(Boolean) : !!value;

const existValuesOnAnyItem = (items: any, fieldPath: string[]) =>
  Object.values(items).some(item => existFieldValue(_.get(item, fieldPath)));

const renderFieldValue = (
  item: any,
  fieldPath: string[],
  fieldPathFallback: string[],
  index: number
) => {
  const fieldValue = _.get(item, fieldPath) || _.get(item, fieldPathFallback);

  return (
    <td key={`row-${fieldPath}-${index}`}>
      {Array.isArray(fieldValue)
        ? fieldValue.map((value, fieldIndex) => (
            <span key={fieldIndex} className='mr-2'>
              {getTranslateText(value)}
            </span>
          ))
        : getTranslateText(fieldValue)}
    </td>
  );
};

const renderRow = (
  label: any,
  items: any,
  fieldPath: string[],
  fieldPathFallback: string[] = []
) =>
  existValuesOnAnyItem(items, fieldPath) ||
  existValuesOnAnyItem(items, fieldPathFallback) ? (
    <tr>
      <td>
        <strong>{label}</strong>
      </td>
      {Object.values(items).map((item, index) =>
        renderFieldValue(item, fieldPath, fieldPathFallback, index)
      )}
    </tr>
  ) : null;

const renderRowUrl = (label: any, items: any, fieldPath: any) =>
  existValuesOnAnyItem(items, fieldPath) ? (
    <tr>
      <td>
        <strong>{label}</strong>
      </td>
      {Object.values(items)
        .map(item => _.get(item, fieldPath))
        .map((fieldValue, index) => (
          <td key={`row-${fieldPath}-${index}`}>
            {_.get(fieldValue, 'uri') ? (
              <LinkExternal href={_.get(fieldValue, 'uri')} external>
                {_.get(fieldValue, 'prefLabel') || _.get(fieldValue, 'uri')}
              </LinkExternal>
            ) : (
              getTranslateText(_.get(fieldValue, 'prefLabel'))
            )}
          </td>
        ))}
    </tr>
  ) : null;

const renderRemoveItem = (
  items: any,
  history: any,
  conceptIdsArray: any[],
  removeConcept: any
) => (
  <tr>
    <td />
    {Object.keys(items).map((item, index) => (
      <td key={`row-button-${index}`}>
        <button
          type='button'
          className='btn fdk-text-size-15 fdk-color-link bg-transparent'
          onClick={() => {
            onDeleteConcept(
              _.get(items, [item, 'id']),
              history,
              conceptIdsArray,
              removeConcept
            );
          }}
        >
          <i className='fa fa-minus-circle' />
          &nbsp;
          {localization.compare.removeCompare}
        </button>
      </td>
    ))}
  </tr>
);

export const ConceptComparePage: FC<Props> = ({
  conceptsCompare,
  fetchConceptsToCompareIfNeeded,
  removeConcept
}) => {
  const history = useHistory();
  const { search } = useLocation();

  const searchParameters = qs.parse(search, {
    ignoreQueryPrefix: true
  });
  const conceptIdsArray = ((searchParameters?.compare ?? '') as string).split(
    ','
  );

  useEffect(() => {
    fetchConceptsToCompareIfNeeded?.(conceptIdsArray);
  }, [searchParameters?.compare]);

  return (
    <main id='content' className='container'>
      <article>
        <div className='row'>
          <div className='col-12'>
            <Helmet>
              <title>{localization.menu.conceptsCompare}</title>
              <meta
                property='og:title'
                content={localization.menu.conceptsCompare}
              />
            </Helmet>
            {conceptsCompare && (
              <>
                <h1 className='title'>
                  {localization.menu.conceptsCompare} (
                  {Object.keys(conceptsCompare).length})
                </h1>

                <section className='scrollable'>
                  <table className='table'>
                    {renderTitle(
                      localization.facet.concept,
                      conceptsCompare,
                      'prefLabel'
                    )}
                    <tbody>
                      {renderRow(
                        localization.responsible,
                        conceptsCompare,
                        ['publisher', 'prefLabel'],
                        ['publisher', 'name']
                      )}
                      {renderRow(
                        localization.compare.definition,
                        conceptsCompare,
                        ['definition', 'text']
                      )}
                      {renderRowUrl(
                        localization.compare.source,
                        conceptsCompare,
                        ['definition', 'source']
                      )}
                      {renderRow(
                        localization.compare.subject,
                        conceptsCompare,
                        ['subject']
                      )}
                      {renderRow(
                        localization.compare.altLabel,
                        conceptsCompare,
                        ['altLabel']
                      )}
                      {renderRow(
                        localization.compare.hiddenLabel,
                        conceptsCompare,
                        ['hiddenLabel']
                      )}

                      {renderRemoveItem(
                        conceptsCompare,
                        history,
                        conceptIdsArray,
                        removeConcept
                      )}
                    </tbody>
                  </table>
                </section>
              </>
            )}
          </div>
        </div>
      </article>
    </main>
  );
};
