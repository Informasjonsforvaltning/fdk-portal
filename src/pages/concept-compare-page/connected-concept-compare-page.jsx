import { connect } from 'react-redux';
import {
  fetchConceptsToCompareIfNeededAction,
  removeConceptAction
} from '../../redux/modules/conceptsCompare';
import { ConceptComparePage } from './concept-compare-page';

const mapStateToProps = ({ conceptsCompare }) => {
  const { items } = conceptsCompare || {
    items: {}
  };

  return {
    conceptsCompare: items
  };
};

const mapDispatchToProps = dispatch => ({
  fetchConceptsToCompareIfNeeded: iDs =>
    dispatch(fetchConceptsToCompareIfNeededAction(iDs)),
  removeConcept: uri => dispatch(removeConceptAction(uri))
});

export const ConnectedConceptComparePage = connect(
  mapStateToProps,
  mapDispatchToProps
)(ConceptComparePage);
