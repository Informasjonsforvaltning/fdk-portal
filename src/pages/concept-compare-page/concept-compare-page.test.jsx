import React from 'react';
import _ from 'lodash';
import { shallow } from 'enzyme';

import { ConceptComparePage } from './concept-compare-page';
import conceptsResponse from './__fixtures/conceptsApiResponse.json';

const conceptItems = _.get(conceptsResponse, ['_embedded', 'concepts']);
const conceptsCompare = _.keyBy(conceptItems, 'id');

test('should render ConceptComparePage correctly with no apiItem', () => {
  const wrapper = shallow(<ConceptComparePage />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ConceptComparePage correctly with conceptItem', () => {
  const wrapper = shallow(
    <ConceptComparePage conceptsCompare={conceptsCompare} />
  );
  expect(wrapper).toMatchSnapshot();
});
