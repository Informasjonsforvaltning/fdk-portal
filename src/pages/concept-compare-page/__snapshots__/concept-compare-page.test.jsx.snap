// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ConceptComparePage correctly with conceptItem 1`] = `
<main
  className="container"
  id="content"
>
  <article>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <SideEffect(DocumentMeta)
          title="Compare concepts"
        />
        <h1
          className="title"
        >
          Compare concepts
           (
          3
          )
        </h1>
        <section
          className="scrollable"
        >
          <table
            className="table"
          >
            <thead
              className="sticky"
            >
              <tr>
                <th>
                  Preferred term
                </th>
                <th
                  key="row-title-prefLabel-0"
                >
                  <h3>
                    engelsk preflabel14
                  </h3>
                </th>
                <th
                  key="row-title-prefLabel-1"
                >
                  <h3>
                    engelsk preflabel19
                  </h3>
                </th>
                <th
                  key="row-title-prefLabel-2"
                >
                  <h3>
                    engelsk preflabel22
                  </h3>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <strong>
                    Responsible
                  </strong>
                </td>
                <td
                  key="row-publisher,prefLabel-0"
                >
                  engelsk pub preflabel14
                </td>
                <td
                  key="row-publisher,prefLabel-1"
                >
                  engelsk pub preflabel19
                </td>
                <td
                  key="row-publisher,prefLabel-2"
                >
                  engelsk pub preflabel22
                </td>
              </tr>
              <tr>
                <td>
                  <strong>
                    Definition
                  </strong>
                </td>
                <td
                  key="row-definition,text-0"
                >
                  engelsk definition14
                </td>
                <td
                  key="row-definition,text-1"
                >
                  engelsk definition19
                </td>
                <td
                  key="row-definition,text-2"
                >
                  engelsk definition22
                </td>
              </tr>
              <tr>
                <td />
                <td
                  key="row-button-0"
                >
                  <button
                    className="btn fdk-text-size-15 fdk-color-link bg-transparent"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fa fa-minus-circle"
                    />
                     
                    Remove from comparison
                  </button>
                </td>
                <td
                  key="row-button-1"
                >
                  <button
                    className="btn fdk-text-size-15 fdk-color-link bg-transparent"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fa fa-minus-circle"
                    />
                     
                    Remove from comparison
                  </button>
                </td>
                <td
                  key="row-button-2"
                >
                  <button
                    className="btn fdk-text-size-15 fdk-color-link bg-transparent"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fa fa-minus-circle"
                    />
                     
                    Remove from comparison
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </section>
      </div>
    </div>
  </article>
</main>
`;

exports[`should render ConceptComparePage correctly with no apiItem 1`] = `
<main
  className="container"
  id="content"
>
  <article>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <SideEffect(DocumentMeta)
          title="Compare concepts"
        />
      </div>
    </div>
  </article>
</main>
`;
